README.md
# ResourcesProvider-ktx
[![Build Status](https://travis-ci.org/GuilhE/ResourcesProvider-ktx.svg?branch=master)](https://travis-ci.org/GuilhE/ResourcesProvider-ktx) [![Android Arsenal](https://img.shields.io/badge/Android%20Arsenal-ResourcesProvider--ktx-brightgreen.svg?style=flat)](https://android-arsenal.com/details/1/7921)

Need a resource but don't remember where to get it from: ContextCompat, ResourcesCompat, AnimationUtils, Context.resources?  
Fear no more, this lib puts them all in one place! 🤩🥳

## Installation

ResourcesProvider-ktx is distributed through [Maven Central](https://search.maven.org/artifact/com.github.guilhe/resources-provider-ktx), [Jcenter](https://bintray.com/gdelgado/android/ResourcesProvider-ktx) and [Jitpack](https://jitpack.io/#GuilhE/ResourcesProvider-ktx).

```groovy
implementation 'com.github.guilhe:resources-provider-ktx:${LATEST_VERSION}'
```
[![Maven Central](https://img.shields.io/maven-central/v/com.github.guilhe/resources-provider-ktx.svg)](https://search.maven.org/search?q=g:com.github.guilhe%20AND%20resources-provider-ktx) [![Download](https://api.bintray.com/packages/gdelgado/android/ResourcesProvider-ktx/images/download.svg)](https://bintray.com/gdelgado/android/ResourcesProvider-ktx/_latestVersion) ![Bintray](https://img.shields.io/bintray/dt/gdelgado/android/ResourcesProvider-ktx)

## Usage

### Resources types

```java
- text(@StringRes resId: Int): CharSequence
- textArray(@ArrayRes resId: Int): Array<CharSequence>
- quantityText(@PluralsRes resId: Int, quantity: Int): CharSequence
- string(@StringRes resId: Int): String
- string(@StringRes resId: Int, vararg formatArgs: Any): String
- stringArray(@ArrayRes resId: Int): Array<String>
- quantityString(@PluralsRes resId: Int, quantity: Int): String
- quantityString(@PluralsRes resId: Int, quantity: Int, vararg formatArgs: Any): String
- integer(@IntegerRes resId: Int): Int
- intArray(@ArrayRes resId: Int): IntArray
- boolean(@BoolRes resId: Int): Boolean
- dimension(@DimenRes resId: Int): Float
- dimensionPixelSize(@DimenRes resId: Int): Int
- dimensionPixelOffset(@DimenRes resId: Int): Int
- drawable(@DrawableRes resId: Int): Drawable?
- drawable(@DrawableRes id: Int, @StyleRes themeResId: Int): Drawable?
- drawableFromAttr(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Drawable?
- drawableForDensity(@DrawableRes id: Int, @StyleRes themeResId: Int, density: Int): Drawable?
- color(@ColorRes resId: Int): Int
- color(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Int
- colorRes(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Int
- colorStateList(@ColorRes resId: Int): ColorStateList?
- colorStateList(@ColorRes id: Int, @StyleRes themeResId: Int): ColorStateList?
- colorStateListFromAttr(@AttrRes attrResId: Int, @StyleRes themeResId: Int): ColorStateList
- font(@FontRes id: Int): Typeface?
- loadAnimation(@AnimRes id: Int): Animation
- resolveAttribute(@AttrRes id: Int, outValue: TypedValue, resolveRefs: Boolean): Boolean
- resolveAttribute(@AttrRes id: Int, @StyleRes themeResId: Int, outValue: TypedValue, resolveRefs: Boolean): Boolean
- value(@DimenRes id: Int, resolveRefs: Boolean): TypedValue
- identifier(name: String, defType: String, defPackage: String): Int
```
### Themes

You can also easily change themed attributes as the following examples:

by `@ColorRes`:  
```java
.setBackgroundColor(resourcesProvider.colorRes(R.attr.colorPrimary, R.style.App_Style_A)
```

by `@ColorInt`:  
```java
.setColor(resourcesProvider.color(R.attr.colorPrimary, R.style.App_Style_B))
```

by `ColorStateList`:  
```java
.backgroundTintList = resourcesProvider.colorStateListFromAttr(R.attr.colorPrimary, R.style.App_Style_C)
.backgroundTintList = resourcesProvider.colorStateList(R.color.color_selector, R.style.App_Style_C)
```

by `drawableFromAttr`:  
```java
.foreground = resourcesProvider.drawableFromAttr(android.R.attr.selectableItemBackground, R.style.App)
```

### Setup

By Kotlin Extensions (View, Activity, Fragment):
```java
class Activity : AppCompatActivity(){
    fun work() {
        resourcesProvider().string(R.string.app_name_label)
    }
}
```
By Dependency Injection (ex: `Dagger`):  
Add to your dependencies graph:
```java
    @JvmStatic
    @Provides
    @Singleton
    fun provideResourcesProvider(context: YourApp): ResourcesProvider{
        return ResourcesProvider(context)
    }
```
Use it:
```java
//by constructor injection
class MyClass @Inject constructor(resourcesProvider: ResourcesProvider)

//by field injection
@Inject
protected lateinit var resourcesProvider: ResourcesProvider

fun work(){
    resourcesProvider.color(R.color.color_1)
}
``` 
Done! 😎
    
## Dependencies
```groovy
'androidx.core:core-ktx'
'androidx.fragment:fragment-ktx'
```

## Bugs and Feedback

For bugs, questions and discussions please use the [Github Issues](https://github.com/GuilhE/ResourcesProvider-ktx/issues).

## LICENSE

Copyright (c) 2019-present GuilhE

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

---

## [1.3.2]

- Repo update

---

## [1.3.1]

- Added drawableFromAttr()
- Repo update

resources-provider-ktx/src/test/java/com/github/guilhe/resourcesprovider/ResourcesProviderTest.kt
package com.github.guilhe.resourcesprovider

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.Color
import android.os.Build
import android.util.TypedValue
import androidx.test.core.app.ApplicationProvider
import com.github.guilhe.resourcesprovider.test.R
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.util.*

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.O_MR1])
class ResourcesProviderTest {

    private lateinit var appContext: Context
    private lateinit var resourcesProvider: ResourcesProvider
    private val ok = "OK"
    private val stringArray = arrayOf("Home", "Work", "Other", "Custom")
    private val wifiPlural = "Wi-Fi networks available"

    @Before
    fun init() {
        appContext = ApplicationProvider.getApplicationContext()
        resourcesProvider = ResourcesProvider(appContext)
    }

    @Test
    fun context() {
        assertEquals(resourcesProvider.ctx, appContext)
    }

    @Test
    fun constructorAlt() {
        assertNotNull(ResourcesProvider(ApplicationProvider.getApplicationContext(), R.style.AppTheme_TextView))
    }

    @Test
    fun text() {
        assertEquals(resourcesProvider.text(R.string.ok), ok)
    }

    @Test
    fun textArray() {
        assertArrayEquals(resourcesProvider.textArray(R.array.emailAddressTypes), stringArray)
    }

    @Test
    fun textQuantity() {
        assertEquals(resourcesProvider.quantityText(R.plurals.wifi_available, 2), wifiPlural)
    }

    @Test
    fun string() {
        assertEquals(resourcesProvider.string(R.string.ok), ok)
    }

    @Test
    fun stringArgs() {
        assertEquals(resourcesProvider.string(R.string.ok, Locale.getDefault()), ok)
    }

    @Test
    fun stringArray() {
        assertArrayEquals(resourcesProvider.stringArray(R.array.emailAddressTypes), stringArray)
    }

    @Test
    fun stringQuantity() {
        assertEquals(resourcesProvider.quantityString(R.plurals.wifi_available, 2), wifiPlural)
    }

    @Test
    fun stringQuantityArgs() {
        assertEquals(resourcesProvider.quantityString(R.plurals.wifi_available, 2, Locale.getDefault()), wifiPlural)
    }

    @Test
    fun int() {
        assertEquals(resourcesProvider.integer(R.integer.status_bar_notification_info_maxnum), 999)
    }

    @Test
    fun intArray() {
        assertArrayEquals(resourcesProvider.intArray(R.array.intArray), intArrayOf(2, 8, 10, 16))
    }

    @Test
    fun bool() {
        assertEquals(resourcesProvider.boolean(R.bool.isTesting), true)
    }

    @Test
    fun dimension() {
        val dp = 10 * Resources.getSystem().displayMetrics.density
        assertEquals(resourcesProvider.dimension(R.dimen.dimen_1), dp)
    }

    @Test
    fun dimensionPixelOffset() {
        val dp = (10 * Resources.getSystem().displayMetrics.density).toInt()
        assertEquals(resourcesProvider.dimensionPixelOffset(R.dimen.dimen_1), dp)
    }

    @Test
    fun dimensionPixelSize() {
        val dp = (10 * Resources.getSystem().displayMetrics.density).toInt()
        assertEquals(resourcesProvider.dimensionPixelSize(R.dimen.dimen_1), dp)
    }

    @Test
    fun drawable() {
        assertNotNull(resourcesProvider.drawable(R.drawable.ic_android_black_24dp))
    }

    @Test
    fun color() {
        assertEquals(resourcesProvider.color(R.color.black), Color.BLACK)
        assertEquals(resourcesProvider.color(android.R.attr.textColor, R.style.AppTheme_TextView), Color.BLACK)
        assertEquals(resourcesProvider.colorRes(android.R.attr.textColor, R.style.AppTheme_TextView), android.R.color.black)
    }

    @Test
    fun colorStateList() {
        var a = resourcesProvider.colorStateList(R.color.color_selector)!!
        var b = ColorStateList(arrayOf(intArrayOf(android.R.attr.state_pressed), intArrayOf()), intArrayOf(Color.BLACK, Color.WHITE))
        assertEquals(
            a.getColorForState(intArrayOf(android.R.attr.state_pressed), 0),
            b.getColorForState(intArrayOf(android.R.attr.state_pressed), 0)
        )

        a = resourcesProvider.colorStateList(R.color.color_selector_attr, R.style.AppTheme_Selector)!!
        b = ColorStateList(arrayOf(intArrayOf(android.R.attr.state_pressed), intArrayOf()), intArrayOf(Color.DKGRAY, Color.WHITE))
        assertEquals(
            a.getColorForState(intArrayOf(android.R.attr.state_pressed), 0),
            b.getColorForState(intArrayOf(android.R.attr.state_pressed), 0)
        )

        a = ColorStateList.valueOf(resourcesProvider.color(android.R.attr.textColor, R.style.AppTheme_TextView))
        assertEquals(a, resourcesProvider.colorStateListFromAttr(android.R.attr.textColor, R.style.AppTheme_TextView))
    }

    @Test
    fun font() {
        assertNotNull(resourcesProvider.font(R.font.roboto_bold))
    }

    @Test
    fun loadAnimation() {
        assertNotNull(resourcesProvider.loadAnimation(R.anim.stay))
    }

    @Test
    fun resolveAttribute() {
        resourcesProvider = ResourcesProvider(appContext, R.style.AppTheme_TextView)
        val value = TypedValue()

        assertEquals(resourcesProvider.resolveAttribute(android.R.attr.actionBarSize, value, false), false)

        assertEquals(resourcesProvider.resolveAttribute(android.R.attr.fontFamily, value, false), true)
        assertEquals(value.type, TypedValue.TYPE_REFERENCE)

        resourcesProvider.resolveAttribute(android.R.attr.textColor, value, true)
        assertEquals(resourcesProvider.color(value.resourceId), resourcesProvider.color(R.color.black))

        val valueStyled = TypedValue()
        resourcesProvider.resolveAttribute(android.R.attr.textColor, R.style.AppTheme_TextView, valueStyled, true)
        assertEquals(valueStyled.resourceId, value.resourceId)
        assertEquals(resourcesProvider.color(valueStyled.resourceId), resourcesProvider.color(value.resourceId))
    }

    @Test
    fun value() {
        assertEquals(resourcesProvider.value(R.bool.isTesting, true).type, TypedValue.TYPE_INT_BOOLEAN)
        assertNotEquals(resourcesProvider.value(R.bool.isTesting, true).type, TypedValue.TYPE_FLOAT)
    }

    @Test
    fun identifier() {
        assertEquals(resourcesProvider.identifier("ic_android_black_24dp", "drawable", appContext.packageName), R.drawable.ic_android_black_24dp)
        assertNotEquals(resourcesProvider.identifier("icon", "drawable", appContext.packageName), R.drawable.ic_android_black_24dp)
    }
}

resources-provider-ktx/src/main/java/com/github/guilhe/resourcesprovider/ResourcesProvider.kt
@file:Suppress("unused", "NOTHING_TO_INLINE")

package com.github.guilhe.resourcesprovider

import android.app.Activity
import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.util.TypedValue
import android.view.ContextThemeWrapper
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.annotation.*
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment

class ResourcesProvider
constructor(val ctx: Context) {

    constructor(ctx: Context, @StyleRes themeResId: Int) : this(ctx) {
        ctx.setTheme(themeResId)
    }

    inline fun text(@StringRes id: Int): CharSequence = ctx.getText(id)

    @Throws(Resources.NotFoundException::class)
    inline fun textArray(@ArrayRes id: Int): Array<CharSequence> = ctx.resources.getTextArray(id)

    @Throws(Resources.NotFoundException::class)
    inline fun quantityText(@PluralsRes id: Int, quantity: Int): CharSequence = ctx.resources.getQuantityText(id, quantity)

    inline fun string(@StringRes id: Int): String = ctx.getString(id)

    inline fun string(@StringRes id: Int, vararg formatArgs: Any): String = ctx.getString(id, *formatArgs)

    @Throws(Resources.NotFoundException::class)
    inline fun stringArray(@ArrayRes id: Int): Array<String> = ctx.resources.getStringArray(id)

    @Throws(Resources.NotFoundException::class)
    inline fun quantityString(@PluralsRes id: Int, quantity: Int): String = ctx.resources.getQuantityString(id, quantity)

    @Throws(Resources.NotFoundException::class)
    inline fun quantityString(@PluralsRes id: Int, quantity: Int, vararg formatArgs: Any): String =
            ctx.resources.getQuantityString(id, quantity, *formatArgs)

    @Throws(Resources.NotFoundException::class)
    inline fun integer(@IntegerRes id: Int): Int = ctx.resources.getInteger(id)

    @Throws(Resources.NotFoundException::class)
    inline fun intArray(@ArrayRes id: Int): IntArray = ctx.resources.getIntArray(id)

    @Throws(Resources.NotFoundException::class)
    inline fun boolean(@BoolRes id: Int): Boolean = ctx.resources.getBoolean(id)

    @Throws(Resources.NotFoundException::class)
    inline fun dimension(@DimenRes id: Int): Float = ctx.resources.getDimension(id)

    @Throws(Resources.NotFoundException::class)
    inline fun dimensionPixelSize(@DimenRes id: Int): Int = ctx.resources.getDimensionPixelSize(id)

    @Throws(Resources.NotFoundException::class)
    inline fun dimensionPixelOffset(@DimenRes id: Int): Int = ctx.resources.getDimensionPixelOffset(id)

    inline fun drawable(@DrawableRes id: Int): Drawable? = ContextCompat.getDrawable(ctx, id)

    inline fun drawable(@DrawableRes id: Int, @StyleRes themeResId: Int): Drawable? =
            ResourcesCompat.getDrawable(ctx.resources, id, ContextThemeWrapper(ctx, themeResId).theme)

    inline fun drawableFromAttr(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Drawable? =
            drawable(TypedValue().also { ContextThemeWrapper(ctx, themeResId).theme.resolveAttribute(attrResId, it, true) }.resourceId)

    inline fun drawableForDensity(@DrawableRes id: Int, @StyleRes themeResId: Int, density: Int): Drawable? =
            ResourcesCompat.getDrawableForDensity(ctx.resources, id, density, ContextThemeWrapper(ctx, themeResId).theme)

    @ColorInt
    inline fun color(@ColorRes id: Int): Int = ContextCompat.getColor(ctx, id)

    @ColorInt
    inline fun color(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Int =
            color(TypedValue().also { ContextThemeWrapper(ctx, themeResId).theme.resolveAttribute(attrResId, it, true) }.resourceId)

    @ColorRes
    inline fun colorRes(@AttrRes attrResId: Int, @StyleRes themeResId: Int): Int =
            TypedValue().also { ContextThemeWrapper(ctx, themeResId).theme.resolveAttribute(attrResId, it, true) }.resourceId

    @Throws(Resources.NotFoundException::class)
    inline fun colorStateList(@ColorRes id: Int): ColorStateList? = ContextCompat.getColorStateList(ctx, id)

    @Throws(Resources.NotFoundException::class)
    inline fun colorStateList(@ColorRes id: Int, @StyleRes themeResId: Int): ColorStateList? =
            ResourcesCompat.getColorStateList(ctx.resources, id, ContextThemeWrapper(ctx, themeResId).theme)

    inline fun colorStateListFromAttr(@AttrRes attrResId: Int, @StyleRes themeResId: Int): ColorStateList =
            ColorStateList.valueOf(color(attrResId, themeResId))

    @Throws(Resources.NotFoundException::class)
    inline fun font(@FontRes id: Int): Typeface? = ResourcesCompat.getFont(ctx, id)

    @Throws(Resources.NotFoundException::class)
    inline fun loadAnimation(@AnimRes id: Int): Animation = AnimationUtils.loadAnimation(ctx, id)

    inline fun resolveAttribute(@AttrRes id: Int, outValue: TypedValue, resolveRefs: Boolean): Boolean =
            ctx.theme.resolveAttribute(id, outValue, resolveRefs)

    inline fun resolveAttribute(@AttrRes id: Int, @StyleRes themeResId: Int, outValue: TypedValue, resolveRefs: Boolean): Boolean =
            ContextThemeWrapper(ctx, themeResId).theme.resolveAttribute(id, outValue, resolveRefs)

    @Throws(Resources.NotFoundException::class)
    inline fun value(@DimenRes id: Int, resolveRefs: Boolean): TypedValue = TypedValue().apply {
        ctx.resources.getValue(id, this, resolveRefs)
    }

    inline fun identifier(name: String, defType: String, defPackage: String): Int = ctx.resources.getIdentifier(name, defType, defPackage)
}

//Helper extensions
fun View.resourcesProvider() = ResourcesProvider(context)

fun View.resourcesProvider(@StyleRes themeResId: Int) = ResourcesProvider(context, themeResId)

fun Activity.resourcesProvider() = ResourcesProvider(this)

fun Activity.resourcesProvider(@StyleRes themeResId: Int) = ResourcesProvider(this, themeResId)

@Throws(IllegalStateException::class)
fun Fragment.resourcesProvider(): ResourcesProvider = ResourcesProvider(requireContext())

@Throws(IllegalStateException::class)
fun Fragment.resourcesProvider(@StyleRes themeResId: Int) = ResourcesProvider(requireContext(), themeResId)


resources-provider-ktx/src/androidTest/java/com/github/guilhe/resourcesprovider/ResourcesProviderTest.kt
package com.github.guilhe.resourcesprovider

import android.content.Context
import android.content.res.Resources
import android.graphics.Color
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.guilhe.resourcesprovider.test.R
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*
import android.content.res.ColorStateList
import android.util.TypedValue

/**
 * Instrumented test, which will execute on an Android device.
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
@RunWith(AndroidJUnit4::class)
class ResourcesProviderTest {

    private lateinit var appContext: Context
    private lateinit var resourcesProvider: ResourcesProvider
    private val ok = "OK"
    private val stringArray = arrayOf("Home", "Work", "Other", "Custom")
    private val wifiPlural = "Wi-Fi networks available"

    @Before
    fun init() {
        appContext = ApplicationProvider.getApplicationContext<Context>()
        resourcesProvider = ResourcesProvider(appContext)
    }

    @Test
    fun context() {
        assertEquals(resourcesProvider.ctx, appContext)
    }

    @Test
    fun text() {
        assertEquals(resourcesProvider.text(R.string.ok), ok)
    }

    @Test
    fun textArray() {
        assertArrayEquals(resourcesProvider.textArray(R.array.emailAddressTypes), stringArray)
    }

    @Test
    fun textQuantity() {
        assertEquals(resourcesProvider.quantityText(R.plurals.wifi_available, 2), wifiPlural)
    }

    @Test
    fun string() {
        assertEquals(resourcesProvider.string(R.string.ok), ok)
    }

    @Test
    fun stringArgs() {
        assertEquals(resourcesProvider.string(R.string.ok, Locale.getDefault()), ok)
    }

    @Test
    fun stringArray() {
        assertArrayEquals(resourcesProvider.stringArray(R.array.emailAddressTypes), stringArray)
    }

    @Test
    fun stringQuantity() {
        assertEquals(resourcesProvider.quantityString(R.plurals.wifi_available, 2), wifiPlural)
    }

    @Test
    fun stringQuantityArgs() {
        assertEquals(resourcesProvider.quantityString(R.plurals.wifi_available, 2, Locale.getDefault()), wifiPlural)
    }

    @Test
    fun int() {
        assertEquals(resourcesProvider.integer(R.integer.status_bar_notification_info_maxnum), 999)
    }

    @Test
    fun intArray() {
        assertArrayEquals(resourcesProvider.intArray(R.array.intArray), intArrayOf(2, 8, 10, 16))
    }

    @Test
    fun bool() {
        assertEquals(resourcesProvider.boolean(R.bool.isTesting), true)
    }

    @Test
    fun dimension() {
        val dp = 10 * Resources.getSystem().displayMetrics.density
        assertEquals(resourcesProvider.dimension(R.dimen.dimen_1), dp)
    }

    @Test
    fun dimensionPixelOffset() {
        val dp = (10 * Resources.getSystem().displayMetrics.density).toInt()
        assertEquals(resourcesProvider.dimensionPixelOffset(R.dimen.dimen_1), dp)
    }

    @Test
    fun dimensionPixelSize() {
        val dp = (10 * Resources.getSystem().displayMetrics.density).toInt()
        assertEquals(resourcesProvider.dimensionPixelSize(R.dimen.dimen_1), dp)
    }

    @Test
    fun drawable() {
        assertNotNull(resourcesProvider.drawable(R.drawable.ic_android_black_24dp))
    }

    @Test
    fun color() {
        assertEquals(resourcesProvider.color(R.color.black), Color.BLACK)
    }

    @Test
    fun colorStateList() {
        val a = resourcesProvider.colorStateList(R.drawable.color_selector)!!
        val b = ColorStateList(
            arrayOf(intArrayOf(android.R.attr.state_pressed), intArrayOf()),
            intArrayOf(Color.BLACK, Color.WHITE)
        )
        assertEquals(a.changingConfigurations, b.changingConfigurations)
    }

    @Test
    fun font() {
        assertEquals(resourcesProvider.font(R.font.roboto_bold)!!.isBold, true)
    }

    @Test
    fun loadAnimation() {
        assertNotNull(resourcesProvider.loadAnimation(R.anim.stay))
    }

    @Test
    fun resolveAttribute() {
        appContext.setTheme(R.style.AppTheme_TextView)
        val value = TypedValue()

        var resolved = resourcesProvider.resolveAttribute(android.R.attr.fontFamily, value, false)
        assertEquals(resolved, true)
        assertEquals(value.type, TypedValue.TYPE_REFERENCE)

        resolved = resourcesProvider.resolveAttribute(android.R.attr.textColor, value, false)
        assertEquals(resolved, true)
        assertEquals(value.type, TypedValue.TYPE_REFERENCE)

        resolved = resourcesProvider.resolveAttribute(android.R.attr.actionBarSize, value, false)
        assertEquals(resolved, false)
    }

    @Test
    fun value() {
        assertEquals(resourcesProvider.value(R.bool.isTesting, true).type, TypedValue.TYPE_INT_BOOLEAN)
        assertNotEquals(resourcesProvider.value(R.bool.isTesting, true).type, TypedValue.TYPE_FLOAT)
    }

    @Test
    fun identifier() {
        assertEquals(resourcesProvider.identifier("ic_android_black_24dp", "drawable", appContext.getPackageName()), R.drawable.ic_android_black_24dp)
        assertNotEquals(resourcesProvider.identifier("icon", "drawable", appContext.getPackageName()), R.drawable.ic_android_black_24dp)
    }
}